### PR TITLE
Tk series page

### PR DESCRIPTION
**Pull request by techncl**
1. Removed the hard-coded asterisk from the series page form label
2. Wrote a function to return a asterisk if input option is required
3. Added 3 tests in order to verify that function (mentioned above) works

### CODE REVIEW

File: public/stylesheets/main.css [hunk 1]
```diff
@@ -0,0 +1 @@
...ast-child,.govuk-radios__item:last-of-type{margin-bottom:0}.govuk-radios__input{cursor:pointer;position:absolute;z-index:1;top:-2px;left:-2px;width:44px;height:44px;margin:0;opacity:0}.govuk-radios__label{display:inline-block;margin-bottom:0;padding:8px 15px 5px;cursor:pointer;-ms-touch-action:manipulation;touch-action:manipulation}.govuk-radios__label::before{content:"";box-sizing:border-box;position:absolute;top:0;left:0;width:40px;height:40px;border:2px solid currentColor;border-radius:50%;background:0 0}.govuk-radios__label::after{content:"";position:absolute;top:10px;left:10px;width:0;height:0;border:10px solid currentColor;border-radius:50%;opacity:0;background:currentColor}.govuk-radios__hint{display:block;padding-right:15px;padding-left:15px}.govuk-radios__input:focus+.govuk-radios__label::before{border-width:4px;box-shadow:0 0 0 4px #fd0}.govuk-radios__input:checked+.govuk-radios__label::after{opacity:1}.govuk-radios__input:disabled,.govuk-radios__input:disabled+.govuk-radios__label{cursor:default}.govuk-radios__input:disabled+.govuk-radios__label{opacity:.5}@media (min-width:40.0625em){.govuk-radios--inline:after{content:"";display:block;clear:both}.govuk-radios--inline .
```
Comment (suzannehamilton) [inline public/stylesheets/main.css]: @Lholbourn and I were discussing this file yesterday - should it be committed? There's going to be a lot of churn, and the Jenkins deployment will have to run npm anyway.

File: test/controllers/SeriesDetailsControllerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import play.api.test.FakeRequest
+import play.api.test.Helpers._
+import util.FrontEndTestHelper
+
+class SeriesDetailsControllerSpec extends FrontEndTestHelper {
+  "SeriesDetailsController GET" should {
+
+    "render the series details page with an authenticated user" in {
+      val controller = new SeriesDetailsController(getAuthorisedSecurityComponents())
+      val home = controller.seriesDetails().apply(FakeRequest(GET, "/"))
```
Comment (suzannehamilton) [inline test/controllers/SeriesDetailsControllerSpec.scala]: Maybe I'm misunderstanding how these controller tests work, but why is this variable called `home`? And why do you make a request to "/" (which looks like the homepage route) to the SeriesDetailsController?

File: test/controllers/SeriesDetailsControllerSpec.scala [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+      status(home) mustBe OK
+      contentType(home) mustBe Some("text/html")
+      contentAsString(home) must include ("seriesDetails.header")
+      contentAsString(home) must include ("seriesDetails.title")
+    }
+
+    "return a redirect to the auth server with an unauthenticated user" in {
+      val controller = new SeriesDetailsController(getUnauthorisedSecurityComponents())
+      val home = controller.seriesDetails().apply(FakeRequest(GET, "/"))
+      redirectLocation(home).get should include ("/auth/realms/tdr/protocol/openid-connect/auth")
```
Comment (suzannehamilton) [inline test/controllers/SeriesDetailsControllerSpec.scala]: Very minor thing, but is there a ScalaTest matcher you can use to avoid calling `.get` (which will cause a confusing error message if it fails)?

File: conf/application.conf [hunk 4]
```diff
@@ -3,7 +3,7 @@ play.http.port=9000
 
 auth.url="http://localhost:8080"
 auth.callback="http://localhost:9000/callback"
-auth.secret="set_me"
+auth.secret="[set me]"
```
Comment (suzannehamilton) [inline conf/application.conf]: I know this isn't really your change, but is it actually necessary to set this secret? I think it gets overridden in all of the deployed environments, and it doesn't matter what you set it to in development.

Maybe setting it to `dev-environment-placeholder` might make it clearer, so that it's obvious that it should only be used in development and you don't have to change it when you set up the project?

File: conf/routes [hunk 5]
```diff
@@ -6,6 +6,8 @@
 # An example controller showing a sample home page
 GET     /                           controllers.HomeController.index
 GET     /dashboard                  controllers.DashboardController.dashboard
+GET     /seriesDetails              controllers.SeriesDetailsController.seriesDetails
+POST    /seriesDetails/submit       controllers.SeriesDetailsController.seriesSubmit()
```
Comment (suzannehamilton) [inline conf/routes]: Not a big deal, but why do we have parentheses in the POST route but not the GET one?

File: app/controllers/SeriesDetailsController.scala [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import play.api.i18n.I18nSupport
+import play.api.mvc.{Action, AnyContent, Request}
+
+@Singleton
+class SeriesDetailsController @Inject()(val controllerComponents: SecurityComponents)
+                                         extends Security[CommonProfile] with I18nSupport  {
+
+  val form = Form(
+    mapping(
+      "seriesNo" -> number
```
Comment (suzannehamilton) [inline app/controllers/SeriesDetailsController.scala]: Maybe call it seriesId or seriesName rather than seriesNo, since it's not a number? How do archivists refer to it?

File: app/controllers/SeriesDetailsController.scala [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    )(SelectedSeriesData.apply)(SelectedSeriesData.unapply)
+  )
+
+  val mockedData: Seq[(String, String)] = Seq(("series1", "Series 1"), ("series2", "Series 2"), ("series3", "Series 3"))
+
+  def seriesDetails(): Action[AnyContent] = Secure("OidcClient") { implicit request: Request[AnyContent] =>
```
Comment (suzannehamilton) [inline app/controllers/SeriesDetailsController.scala]: Is there a way of commonising this `Secure("OidcClient")` wrapper so we don't have to set a String every time we add  a new controller action?

File: app/views/inputDropdown.scala.html [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    <label class="govuk-label" for="@elements.id">
+        @elements.args.get('_label)
+        @elements.args.requiredLabelSuffix()
+    </label>
+    <select
+            class="govuk-select"
+            id="@elements.id"
+            name="@elements.id"
+            @elements.args.requiredInput()>
+        <option value="" selected>Please choose...</option>
```
Comment (suzannehamilton) [inline app/views/inputDropdown.scala.html]: Did Pia have wording for this on her wireframes? I'm just wondering if something like "Select a series" might be more direct.

File: app/views/inputDropdown.scala.html [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+        @for(dropdownOption <- dropdownOptions) {
+            <option value="@dropdownOption._1">@dropdownOption._2</option>
```
Comment (suzannehamilton) [inline app/views/inputDropdown.scala.html]: Are you planning to change this to a case class when we have real data? Tuples are generally a bit hard to follow.

File: app/views/inputDropdown.scala.html (hunk 8 above)
Comment (TomJKing) [reply app/views/inputDropdown.scala.html]: This is the wording that is used on the wireframes: https://projects.invisionapp.com/share/S4TVATFN82M#/screens/383373576

File: conf/application.conf (hunk 4 above)
Comment (TomJKing) [reply conf/application.conf]: In the local environment it needs to be set to the tdr client crendentials secret that is setup in the local keycloak application.

File: app/views/inputDropdown.scala.html (hunk 9 above)
Comment (TomJKing) [reply app/views/inputDropdown.scala.html]: Yes. At the moment not clear what the data will be. Added comment to the template about this.

File: conf/application.conf (hunk 4 above)
Comment (suzannehamilton) [reply conf/application.conf]: Should it be read from an environment variable, then? Otherwise (as far as I can tell) you'll have changes that you have to avoid committing.

File: app/views/inputDropdown.scala.html (hunk 8 above)
Comment (suzannehamilton) [reply app/views/inputDropdown.scala.html]: OK, let's leave it then. It just sounds odd to me. :slightly_smiling_face: 

And I can't find any examples in GOV.UK because [they usually prefer radio buttons](https://design-system.service.gov.uk/components/select/), which might not be appropriate if we have to show any long series lists.

File: conf/routes (hunk 5 above)
[the remark this answers is not in the document]
Comment (TomJKing) [reply conf/routes]: What path scheme do you think we should adopt?

File: conf/messages [hunk 10]
```diff
@@ -7,4 +7,7 @@ home.use.transfer.resume=resume a transfer
 home.use.transfer.history=view your transfer history
 home.start=Start now
 dashboard.header=Dashboard
-dashboard.title=Welcome to the dashboard
\ No newline at end of file
+dashboard.title=Welcome to the dashboard
+seriesDetails.header=SeriesDetails
```
[the remark this answers is not in the document]
Comment (TomJKing) [reply conf/messages]: Resolved

File: test/controllers/SeriesDetailsControllerSpec.scala (hunk 2 above)
Comment (TomJKing) [reply test/controllers/SeriesDetailsControllerSpec.scala]: Resolved

File: app/controllers/SeriesDetailsController.scala (hunk 7 above)
Comment (TomJKing) [reply app/controllers/SeriesDetailsController.scala]: Tried to do something here but not sure if its any good

File: conf/routes (hunk 5 above)
[the remark this answers is not in the document]
Comment (suzannehamilton) [reply conf/routes]: Just discussed in person. Let's change it to this for now:

```
GET   /series
POST  /series
```

Could you put a card in the tech backlog so we can plan the routes soon?

File: conf/application.conf [hunk 11]
```diff
@@ -3,7 +3,8 @@ play.http.port=9000
 
 auth.url="http://localhost:8080"
 auth.callback="http://localhost:9000/callback"
-auth.secret="set_me"
+auth.secret="secretValue"
+auth.secret=${?AUTH_SECRET}
```
Comment (suzannehamilton) [inline conf/application.conf]: What will happen if you forget to set the environment variable? Is the error obvious, or does login just mysteriously fail?

File: app/controllers/SeriesDetailsController.scala [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+@Singleton
+class SeriesDetailsController @Inject()(val controllerComponents: SecurityComponents)
+                                         extends Security[CommonProfile] with I18nSupport  {
+
+  private val secureAction = Secure("OidcClient")
+
+  val selectedSeriesForm = Form(
+    mapping(
+      "seriesId" -> number
```
Comment (suzannehamilton) [inline app/controllers/SeriesDetailsController.scala]: How does `number` work here, given that the mock IDs are strings like `mockedSeries1`?

File: app/controllers/SeriesDetailsController.scala (hunk 7 above)
Comment (suzannehamilton) [reply app/controllers/SeriesDetailsController.scala]: Yep, that's the kind of thing I had in mind (assuming it's fine to use the same `Secure` instance for each request). You could also move it to its own file so it can be reused in other controllers, or we can do that later.

File: conf/application.conf (hunk 11 above)
Comment (TomJKing) [reply conf/application.conf]: Error: [CompletionException: org.pac4j.core.exception.TechnicalException: Bad token response, error=unauthorized_client]

File: conf/routes (hunk 5 above)
[the remark this answers is not in the document]
Comment (TomJKing) [reply conf/routes]: Card added and routes changed

File: conf/application.conf (hunk 11 above)
Comment (suzannehamilton) [reply conf/application.conf]: That's probably good enough for now. If it trips people up, we can put some more checks around it.

File: conf/routes (hunk 5 above)
[the remark this answers is not in the document]
Comment (suzannehamilton) [reply conf/routes]: :+1:
